Initial commit of the approval resource

Seraf opened the pull request.

Seraf committed:
Initial commit

Seraf committed:
Merge branch 'master' into develop

Seraf committed:
logs: Adding some logs on lock and release

There was no output when the resource was acquiring or releasing a lock. This commit output as INFO log these informations

Seraf committed:
docs: Adding documentation on code and Readme

This commit explains how the code work by adding some useful comments and add a README to explain how to configure and use it

xlr-8 committed:
README: Fixing typos

There were couple minor typos in the text that have been fixed.

xlr-8 commented on Dockerfile:
```diff
@@ -0,0 +1,15 @@
+
+# test
+ADD test/ /opt/resource-tests/
+#RUN /opt/resource-tests/test.sh
```
Is the comment still necessary?

xlr-8 commented on README.md:
````diff
@@ -1 +1,164 @@
+```
+# python3 cli.py reject --id f47f5864-a1b4-4bc0-8899-784ae3b768a4
+INFO: The lock f47f5864-a1b4-4bc0-8899-784ae3b768a4 has been rejected
+```
````
Nice README!
Could eventually add a license I guess?

Seraf replied:
Already done in https://github.com/cycloidio/approval-resource/blob/master/LICENSE

xlr-8 commented on ci/pipeline.yml:
```diff
@@ -0,0 +1,119 @@
+        trigger: true
+        passed: [push-dev-image]
+      #- task: test-unit
+      #  file: resource-src/ci/tasks/test-unit.yml
```
Leaving comments?
Couple more further down as well.

Seraf replied:
Yes, I'm leaving them, I would like to add unit test in the future

xlr-8 reviewed (commented):
I have done a very brief review, it looks pretty good. I commented minor things, haven't gone deeply into the code.

gaelL commented on assets/approval.py:
```diff
@@ -0,0 +1,344 @@
+
+            # We want to wait until the approve is done
+            if approval_lock.need_approval:
+                while approval_lock.approved is None:
```
Maybe useless if statement.

We can use it directly on the while with both condition.

`while approval_lock.need_approval and approval_lock.approved is None`

gaelL replied:
To clarify if statements maybe do something like : 

```
        while approval_lock.approved is None and approval_lock.need_approval:
            log.info("The lock %s is waiting for an approval" % params['lock_name'])
            # Query the lock item in the loop
            refresh_approval = self.query_lock(params['lock_name'])

            # If hasn't been approved or rejected, waiting a bit more
            if refresh_approval.approved is none:
                time.sleep(self.wait_lock)
                continue

            # Is it approved ?
            if refresh_approval.approved:
                approval_lock.approved = True
            # If not, we should fail the job and release the lock
            else:
                log.info("The lock hasn't been approved, exiting")
                approval_lock.claimed = False
                approval_lock.approved = None
                self.engine.save(approval_lock, overwrite=True)
                exit(1)
```

Seraf replied:
:+1:

gaelL commented on assets/approval.py:
```diff
@@ -0,0 +1,344 @@
+                .first(desc=True)
+
+        metadata = []
+        if approval_lock:
```
Replace directly the if by a if not to remove the else statement.
```
if not approval_lock:
    log.info("No lock have been found")
    exit(0)
```

Seraf replied:
:+1:

gaelL commented on assets/approval.py:
```diff
@@ -0,0 +1,344 @@
+                value = getattr(approval_lock, key)
+                if type(value) is datetime:
+                    value = str(Decimal(value.timestamp()))
+                if type(value) is bool:
```
maybe `elif` to avoid to evaluate the statement again in case of the first one match

Seraf replied:
:+1:

gaelL commented on assets/approval.py:
```diff
@@ -0,0 +1,344 @@
+        if 'claim' in params['action']:
+            if approval_lock:
+                # If the lock is claimed in database
+                if approval_lock.claimed:
```
Same as the comment on line 134

gaelL commented on assets/approval.py:
```diff
@@ -0,0 +1,344 @@
+                self.engine.save(approval_lock, overwrite=True)
+                log.info("Releasing the lock %s" % params['lock_name'])
+
+            else:
```
Same as line 161

gaelL commented on assets/approval.py:
```diff
@@ -0,0 +1,344 @@
+        metadata = []
+        for key in approval_lock.keys_():
+            value = getattr(approval_lock, key)
+            if type(value) is datetime:
```
Same as line 165

gaelL commented on assets/approval.py:
```diff
@@ -0,0 +1,344 @@
+        approval_lock = self.query_lock(params['lock_name'])
+        need_approval = params.get('need_approval', False)
+
+        if 'claim' in params['action']:
```
WDYT about split actions in private functions like `_do_claim` and `_do_release` ?

gaelL commented on assets/approval.py:
```diff
@@ -0,0 +1,344 @@
+
+        return json.dumps(response)
+
+print(ApprovalResource(os.path.basename(__file__), sys.stdin.read(), sys.argv[1:]).run())
```
Regarding python best practices for a script I would add `if __name__ == "__main__":` for this line.

And also used named arguments to be more readable like `info(spacing=10, collapse=1):`

Seraf committed:
beautify: More pythonic code

Seraf committed:
approval: Adding an override option

This commit add an override option to let the put action force the last approval to fail by rejecting it.
It will then be able to claim the lock.

The usecase is to fail the previous pipeline with the old version to approve only the new one

Seraf committed:
timeout: Adding timeout on the approval

If the approval is too long, the resource will automatically reject the current approval. The timeout is defined in minutes and is passed as parameter for the get

Seraf committed:
ci: Use the master branch to build the docker image
